lib/logster/redis_store.rb

```diff
@@ -2,6 +2,80 @@
 require 'logster/base_store'
 
 module Logster
+  class RedisRateLimiter
+    BUCKETS = 6
+
+    attr_reader :key, :set_key, :callback_key
+
+    def initialize(redis, severities, limit, duration, block = nil)
+      @redis = use_raw_connection? ? Logster.config.redis_raw_connection : redis
+      @severities = severities
+      @limit = limit
+      @duration = duration
+      @block = block
+
+      # "_LOGSTER_RATE_LIMIT:012:20:30"
+      # Triggers callback when log levels of :debug, :info and :warn occurs 20 times within 30 secs
+      @key = "#{key_prefix}#{@severities.join("")}:#{@limit}:#{@duration}"
+      @set_key = "#{@key}:set"
+      @callback_key = "#{@key}:callback_triggered"
+      @bucket_range = @duration / BUCKETS
+    end
+
+    def perform(severity)
+      return unless @severities.include?(severity)
+      time = Time.now.to_i
+      redis_key = "#{@key}:#{bucket_number(time)}"
+      current_rate = retrieve_rate
```

Comment:
this pattern is all a bit off imo, you should not need to call retrieve_rate every time... instead have the lua routine do absolutely everything ... so when it is done with the INCR have it count all the other buckets.